src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs

```diff
@@ -1337,4 +1357,156 @@ internal bool CanAddValue(HttpHeaderParser parser)
+                {
+                    if (value is HttpHeaders.HeaderStoreItemInfo info)
+                    {
+                        if (info.OriginalRawValue != null)
+                        {
+                            return info.OriginalRawValue is string rawString ? new HeaderStringValues(rawString) : new HeaderStringValues((List<string>)info.OriginalRawValue);
+                        }
+
+                        return new HeaderStringValues(HttpHeaders.GetValuesAsStrings(descriptor, info));
+                    }
+
+                    Debug.Assert(value is string);
+                    return new HeaderStringValues((string)value);
+                }
+            }
+
+            public KeyValuePair<string, HeaderStringValues> Current { get; private set; }
+
+            public void Dispose()
+            {
+                _headerStore.Dispose();
+            }
+
+            object? IEnumerator.Current => Current;
+
+            void IEnumerator.Reset()
+            {
+                _headerStore.Reset();
+            }
+        }
+    }
+
+    public readonly struct HeaderStringValues : IEnumerable<string>
+    {
+        private readonly IEnumerable<string?>? _values;
+        private readonly string? _value;
+
+        public HeaderStringValues(string value)
+        {
+            _value = value;
+            _values = null;
+        }
+
+        public HeaderStringValues(IEnumerable<string?> value)
+        {
+            _values = value;
+            _value = null;
+        }
+
+        public Enumerator GetEnumerator() => new Enumerator(_values, _value);
+
+        IEnumerator<string> IEnumerable<string>.GetEnumerator() => GetEnumerator();
+
+        IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+
+        public struct Enumerator : IEnumerator<string>
+        {
+            private readonly IEnumerator<string?>? _values;
+            private readonly string? _value;
+            private bool _completed;
+
+            public Enumerator(IEnumerable<string?>? values, string? value)
+            {
+                _values = values?.GetEnumerator();
+                _value = value;
+                _completed = false;
+                Current = default!;
+            }
+
+            public bool MoveNext()
+            {
+                if (_completed)
+                {
+                    return false;
+                }
+
+                if (_values == null)
+                {
+                    _completed = true;
+                    Current = _value!;
+                    return true;
+                }
+
+                _completed = !_values.MoveNext();
+                if (_completed)
+                {
+                    return false;
+                }
+
+                Debug.Assert(_values.Current != null);
+                Current = _values.Current!;
+
+                return true;
+            }
+
+            public string Current { get; private set; }
```

Comment:
The same thing with the setter on interface implementation as above.

Reply:
Discussed above.